### PR TITLE
leds: qcom-lpg: Add support for PMI8950 PWM

### PR DESCRIPTION
**Pull request by M0Rf30**
… markw

Need to set:

CONFIG_RC_CORE=m
CONFIG_RC_MAP=m
CONFIG_LIRC=y
CONFIG_LEDS_GPIO=m
CONFIG_IR_GPIO_CIR=m
CONFIG_IR_GPIO_TX=m
CONFIG_IR_PWM_TX=m

### CODE REVIEW

**Comment by vipaoL** [comment]
I can confirm it works on markw, I can turn the ir LED on by executing echo 1 > /sys/class/leds/irled/brightness

**Comment by M0Rf30** [comment]
@kiciuk @marc0x1 can you give feedback for mido?

File: arch/arm64/boot/dts/qcom/msm8953-xiaomi-mido.dts [hunk 1]
```diff
@@ -49,6 +49,20 @@
...
 
+	gpio-leds {
+		compatible = "gpio-leds";
+		pinctrl-0 = <&irled_default>;
+		pinctrl-names = "default";
+
+		irda-led {
+			label = "irled";
+			gpios = <&tlmm 45 GPIO_ACTIVE_HIGH>;
+			linux,default-trigger = "infra-red";
```
Comment (z3ntu) [inline arch/arm64/boot/dts/qcom/msm8953-xiaomi-mido.dts:60]: There's no trigger under that name

**Comment by z3ntu** [comment]
Is there no upstream driver for GPIO IR LED like with `ir-spi-led`? While I haven't looked into the `ir-spi-led` driver I doubt it's getting exposed as a simple LED device, I'm guessing there's some kernel API for IR that then user space uses?

**Comment by M0Rf30** [comment]
> Is there no upstream driver for GPIO IR LED like with `ir-spi-led`? While I haven't looked into the `ir-spi-led` driver I doubt it's getting exposed as a simple LED device, I'm guessing there's some kernel API for IR that then user space uses?

Yeah maybe I found gpio-ir-tx. I will give it a try

**Comment by marc0x1** [comment]
> @Kiciuk @marc0x1 can you give feedback for mido?
Can you provide, if you have it, the compiled version?

**Comment by barni2000** [comment]
I have tested on mido the driver is initialized and it creates /dev/lirc0. I try to use with ir-ctl but unfortunately it doesn't work for me.

**Comment by barni2000** [comment]
As i see in the downstream source for mido we should use [pwm-ir-tx](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/leds/irled/pwm-ir-tx.yaml?h=v6.4)

Downstream node:
```
pwm_ir {
	compatible = "pwm-ir";
	pwms = <&pmi8950_pwm 0 0>;
	reg-id = "vdd";
	vdd-supply = <&pm8953_l8>;
};
```

**Comment by M0Rf30** [comment]
> As i see in the downstream source for mido we should use [pwm-ir-tx](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/leds/irled/pwm-ir-tx.yaml?h=v6.4)
> 
> Downstream node:
> 
> ```
> pwm_ir {
> 	compatible = "pwm-ir";
> 	pwms = <&pmi8950_pwm 0 0>;
> 	reg-id = "vdd";
> 	vdd-supply = <&pm8953_l8>;
> };
> ```

Thanks for the suggestion. I will test today I think

**Comment by M0Rf30** [comment]
> compatible = "pwm-ir";

I enabled the pwm generator for pmi8950 and use of pwm-ir-tx,
I can't test it. let me know

**Comment by barni2000** [comment]
@M0Rf30 I am going to check on mido, but i see you changed markw but it uses gpio-ir. The two device has different irled solution.

**Comment by barni2000** [comment]
> @barni2000 @vipaoL I enabled the pwm generator for pmi8950 and use of pwm-ir-tx, I can't test it. let me know Judging from various dts the pwm node should live in pmic@3 node, has a reg = 0xb000 and one channel (`chanid=0`). So basically a msm8916-pwm with different .base. Some devices (sdm450) also enable a pm8950_pwm (not pmi), on pmic@1 and reg 0xbc00 (seems equivalent to msm8916-pwm)
> 
> Don't forget to enable required kernel modules
> 
> https://github.com/search?q=repo%3ASonicBSV%2Fandroid_kernel_xiaomi_markw-4.9%20pmi8950_pwm&type=code

I have test but unfortunately it is not reacts for anything. The drivers are initialized i have /dev/lirc0 but the led is not working for me.

**Comment by barni2000** [comment]
I have added pmi8950_gpio pinctrl for mido and it is working now.

File: drivers/leds/rgb/leds-qcom-lpg.c [hunk 2]
```diff
@@ -1338,6 +1338,13 @@ static const struct lpg_data pm8941_lpg_data = {
 	},
 };
 
+static const struct lpg_data pm8950_pwm_data = {
```
Comment (barni2000) [inline drivers/leds/rgb/leds-qcom-lpg.c:1341]: it should be called as pmi8950 instead of pm8950.

File: drivers/leds/rgb/leds-qcom-lpg.c [hunk 3]
```diff
@@ -1449,6 +1456,7 @@ static const struct of_device_id lpg_of_table[] = {
 	{ .compatible = "qcom,pm8350c-pwm", .data = &pm8350c_pwm_data },
 	{ .compatible = "qcom,pm8916-pwm", .data = &pm8916_pwm_data },
 	{ .compatible = "qcom,pm8941-lpg", .data = &pm8941_lpg_data },
+	{ .compatible = "qcom,pm8950-pwm", .data = &pm8950_pwm_data },
```
Comment (barni2000) [inline drivers/leds/rgb/leds-qcom-lpg.c:1459]: pmi8950

**Review by barni2000 — CHANGES_REQUESTED** [review]

**Comment by barni2000** [comment]
@M0Rf30 will you send it to upstream?

**Comment by M0Rf30** [comment]
I think to upstream this and posting a link here

**Comment by M0Rf30** [comment]
In linux-next
https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/?h=next-20240415&id=2573c25e2c482b53b6e1142ff3cd28f6de13e659
https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/?h=next-20240415&id=fc3b23faa14371182ca8a3806854f0db827cb91f

**Pull request closed without merging**